tests/run_data_utils_test.py

```diff
@@ -40,7 +40,27 @@ def test_constraint_system_list(self):
                 [ 0.,  1.,  0.,  0.],
                 [ 0.,  0.,  1.,  0.],
                 [ 0.,  0.,  0.,  1.]])))
-        assert np.all(np.equal(dmat, np. array([[1., 0., 1., 0., 0.],
+        assert np.all(np.equal(dmat, np.array([[1., 0., 1., 0., 0.],
+                [0., 1., 0., 1., 0.],
+                [0., 1., 0., 1., 0.],
+                [1., 0., 0., 0., 0.],
+                [1., 0., 0., 0., 0.],
+                [0., 1., 0., 0., 1.]])))
+    
+    def test_constraint_system_list(self):
+        formula = '~0 + condition + batch'
+        sample_description = pd.DataFrame({ 'condition': [0, 1, 1, 0, 0, 1], 'batch': [1, 2, 2, 0, 0, 3] })
+        dmat = pd.DataFrame({ 'condition_0': [1, 0, 0, 1, 1, 0], 'condition_1': [0, 1, 1, 0, 0, 1], 'batch_0': [0, 0, 0, 1, 1, 0], 'batch_1': [1, 0, 0, 0, 0, 0], 'batch_2': [0, 1, 1, 0, 0, 0], 'batch_3': [0, 0, 0, 0, 0, 1] })
+        constraints = np.array([[-1, 0, 0, 0, 0], [0, 1, 0, 0, 0], [0, 0, 1, 0, 0], [0, 0, 0, 1, 0], [0, 0, 0, 0, 1]])
```

Comment:
This cannot work because the rank of dmat is equal to the number of unique rows in our case since we have one hot encoded dmat which guarantees that no row can be a linear combination of any other rows. Therefore you can guess the number of required coeffs in your constraints matrix by counting unique rows. This is 4 in this case but you supply 5 columns @ilan-gold . I guess what you wanted to do is to collapse one coefficient of dmat into an intercept? But you need to collapse 2, namely the first batch + condition like it is done in the other tests in one or the other way. You need to combine the first two column in your supplied cmat. This will result in `np.matmul(dmat, cmat)`:
```py
[[-1.  1.  0.  0.]
 [ 1.  0.  1.  0.]
 [ 1.  0.  1.  0.]
 [-1.  0.  0.  0.]
 [-1.  0.  0.  0.]
 [ 1.  0.  0.  1.]]
```
This is of course equivalent to test_contraint_system_list. I guess supplying a constraints matrix can be checked rather easily by checking if the number of coefficients in constraints is equal to the rank of `np.matmul(dmat, cmat)` which we do. It just doesn't make sense to have a cmat with more coeffs than there are unique samples in dmat.